File: src/gtm_sourcing_agent/stages/search_strategy.py
"""Stage 6: Search Strategy generation. Regenerable independently of
target companies / title intelligence. See prompts/search_strategy.md."""

import json

from .. import llm_client, storage
from ..models import TalentMap


def run(role_id: str, *, storage_backend=storage) -> TalentMap:
    talent_map = storage_backend.require_section(role_id, "talent_map")
    prompt = llm_client.render_prompt("search_strategy.md", talent_map_json=json.dumps(talent_map))
    result = llm_client.generate(prompt, TalentMap, stage="search_strategy")

    existing = storage_backend.load_role(role_id)["talent_map"]
    merged = TalentMap(
        target_companies=existing.get("target_companies", []),
        title_intelligence=existing.get("title_intelligence", {}),
        search_strategies=result.search_strategies,
    )
    storage_backend.merge_section(role_id, "talent_map", merged.model_dump())
    return merged

File: backend-node/src/stages/searchStrategy.ts
// Port of stages/search_strategy.py -- Stage 6: Search Strategy
// generation. Regenerable independently of target companies / title
// intelligence.
import * as storage from "../db/storage.js";
import * as llmClient from "../llmClient.js";
import { TalentMap } from "../models.js";

export async function run(roleId: string): Promise<TalentMap> {
  const talentMap = await storage.requireSection(roleId, "talent_map");
  const prompt = llmClient.renderPrompt("search_strategy.md", { talent_map_json: JSON.stringify(talentMap) });
  const result = await llmClient.generate(prompt, TalentMap, { stage: "search_strategy" });

  const existing = (await storage.loadRole(roleId)).talent_map;
  const merged: TalentMap = {
    target_companies: existing?.target_companies ?? [],
    title_intelligence: existing?.title_intelligence ?? {
      exact_target_titles: [], alternative_titles: [], previous_titles: [], adjacent_titles: [],
      market_terminology: [], competitor_titles: [], geography_specific_titles: [],
    },
    search_strategies: result.search_strategies,
  };
  await storage.mergeSection(roleId, "talent_map", merged);
  return merged;
}
